File: 3501.maximize-active-section-with-trade-ii.go
#
# @lc app=leetcode id=3501 lang=golang
#
# [3501] Maximize Active Section with Trade II
#
# @lc code=start
func maxActiveSectionsAfterTrade(s string, queries [][]int) []int {
	const LOG = 17
	const MAXK = 50005
	n := len(s)
	totalOnes := 0
	for _, ch := range s {
		if ch == '1' {
			totalOnes++
		}
	}

	islandsSt := []int{}
	islandsEn := []int{}
	i := 0
	for i < n {
		if s[i] == '1' {
			st := i
			for i < n && s[i] == '1' {
				i++
			}
			en := i - 1
			islandsSt = append(islandsSt, st)
			islandsEn = append(islandsEn, en)
		} else {
			i++
		}
	}
	k := len(islandsSt)

	fullAdj := make([]int, k)
	for j := 0; j < k; j++ {
		prevB := 0
		if j > 0 {
			prevB = islandsEn[j-1] + 1
		}
		leftF := islandsSt[j] - prevB
		nextB := n
		if j+1 < k {
			nextB = islandsSt[j+1]
		}
		rightF := nextB - 1 - islandsEn[j]
		fullAdj[j] = leftF + rightF
	}

	// Sparse table
	st := make([][]int, LOG)
	logg := make([]int, n+2)
	logg[1] = 0
	for i := 2; i <= n; i++ {
		logg[i] = logg[i/2] + 1
	}
	for lv := 0; lv < LOG; lv++ {
		st[lv] = make([]int, k)
	}
	copy(st[0], fullAdj)
	for lv := 1; lv < LOG; lv++ {
		for i := 0; i+(1<<lv) <= k; i++ {
			st[lv][i] = maxInt(st[lv-1][i], st[lv-1][i+(1<<(lv-1))])
		}
	}

	qrmq := func(L, R int) int {
		if L > R {
			return 0
		}
		leng := R - L + 1
		lv := logg[leng]
		return maxInt(st[lv][L], st[lv][R-(1<<lv)+1])
	}

	// Binary search helpers
	lowerBound := func(arr []int, val int) int {
		lo := 0
		hi := k
		for lo < hi {
			mid := lo + (hi-lo)/2
			if arr[mid] >= val {
				hi = mid
			} else {
				lo = mid + 1
			}
		}
		return lo
	}

	ans := make([]int, len(queries))
	for idx, q := range queries {
		l, r := q[0], q[1]
		jLeft := lowerBound(islandsSt, l+1)
		jRight := lowerBound(islandsEn, r) - 1
		if jLeft > jRight {
			ans[idx] = totalOnes
			continue
		}

		maxG := 0
		// First: jLeft
		{
			prevB := 0
			if jLeft > 0 {
				prevB = islandsEn[jLeft-1] + 1
			}
			leftH := islandsSt[jLeft] - maxInt(l, prevB)
			nextB := n
			if jLeft+1 < k {
				nextB = islandsSt[jLeft+1]
			}
			rightH := minInt(r, nextB-1) - islandsEn[jLeft]
			maxG = maxInt(maxG, leftH+rightH)
		}
		// Last: jRight
		if jRight >= jLeft {
			prevB := 0
			if jRight > 0 {
				prevB = islandsEn[jRight-1] + 1
			}
			leftH := islandsSt[jRight] - maxInt(l, prevB)
			nextB := n
			if jRight+1 < k {
				nextB = islandsSt[jRight+1]
			}
			rightH := minInt(r, nextB-1) - islandsEn[jRight]
			maxG = maxInt(maxG, leftH+rightH)
		}
		// Middles
		if jLeft+1 <= jRight-1 {
			midMax := qrmq(jLeft+1, jRight-1)
			maxG = maxInt(maxG, midMax)
		}
		ans[idx] = totalOnes + maxG
	}
	return ans
}

func maxInt(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func minInt(a, b int) int {
	if a < b {
		return a
	}
	return b
}
# @lc code=end

File: 3501.maximize-active-section-with-trade-ii.java
#
# @lc app=leetcode id=3501 lang=java
#
# [3501] Maximize Active Section with Trade II
#

# @lc code=start
class Solution {
    public List<Integer> maxActiveSectionsAfterTrade(String s, int[][] queries) {
        int n = s.length();
        int total_ones = 0;
        for (int i = 0; i < n; i++) {
            if (s.charAt(i) == '1') total_ones++;
        }
        java.util.ArrayList<Integer> ps_ = new java.util.ArrayList<>();
        java.util.ArrayList<Integer> qs_ = new java.util.ArrayList<>();
        java.util.ArrayList<Integer> ls_ = new java.util.ArrayList<>();
        java.util.ArrayList<Integer> rs_ = new java.util.ArrayList<>();
        for (int i = 0; i < n; ) {
            if (i > 0 && s.charAt(i) == '1' && s.charAt(i - 1) == '0') {
                int p = i;
                int q = i;
                while (q + 1 < n && s.charAt(q + 1) == '1') q++;
                if (q + 1 < n && s.charAt(q + 1) == '0') {
                    // left_len
                    int ll = 0;
                    for (int pos = p - 1; pos >= 0 && s.charAt(pos) == '0'; pos--) ll++;
                    // right_len
                    int rr = 0;
                    for (int pos = q + 1; pos < n && s.charAt(pos) == '0'; pos++) rr++;
                    ps_.add(p);
                    qs_.add(q);
                    ls_.add(ll);
                    rs_.add(rr);
                }
                i = q + 1;
            } else {
                i++;
            }
        }
        int m = ps_.size();
        if (m == 0) {
            java.util.List<Integer> answer = new java.util.ArrayList<>();
            for (int[] qu : queries) {
                answer.add(total_ones);
            }
            return answer;
        }
        int[] Ps = new int[m];
        int[] Qs = new int[m];
        int[] Ls = new int[m];
        int[] Rs = new int[m];
        int[] Vals = new int[m];
        for (int j = 0; j < m; j++) {
            Ps[j] = ps_.get(j);
            Qs[j] = qs_.get(j);
            Ls[j] = ls_.get(j);
            Rs[j] = rs_.get(j);
            Vals[j] = Ls[j] + Rs[j];
        }
        // sparse table
        int LOG = 18;
        int[][] sparse = new int[LOG][m];
        for (int j = 0; j < m; j++) {
            sparse[0][j] = Vals[j];
        }
        for (int k = 1; k < LOG; k++) {
            for (int j = 0; j + (1 << k) <= m; j++) {
                sparse[k][j] = Math.max(sparse[k - 1][j], sparse[k - 1][j + (1 << (k - 1))]);
            }
        }
        java.util.List<Integer> answer = new java.util.ArrayList<>();
        for (int[] qu : queries) {
            int l = qu[0];
            int r = qu[1];
            // jL: first Ps[j] > l
            int jL = 0;
            int jH = m;
            while (jL < jH) {
                int mid = (jL + jH) / 2;
                if (Ps[mid] > l) {
                    jH = mid;
                } else {
                    jL = mid + 1;
                }
            }
            // jR: first Qs[j] >= r , -1
            int low = 0;
            int hi = m;
            while (low < hi) {
                int mid = (low + hi) / 2;
                if (Qs[mid] >= r) {
                    hi = mid;
                } else {
                    low = mid + 1;
                }
            }
            int jR = low - 1;
            if (jL > jR) {
                answer.add(total_ones);
                continue;
            }
            int numc = jR - jL + 1;
            int mg = 0;
            if (numc <= 2) {
                for (int jj = jL; jj <= jR; jj++) {
                    int pj = Ps[jj];
                    int qj = Qs[jj];
                    int ljv = Ls[jj];
                    int rjv = Rs[jj];
                    int le = Math.min(pj - l, ljv);
                    int re = Math.min(r - qj, rjv);
                    mg = Math.max(mg, le + re);
                }
            } else {
                int lefti = jL + 1;
                int righti = jR - 1;
                int klen = righti - lefti + 1;
                int kk = 31 - Integer.numberOfLeadingZeros(klen);
                int imax = Math.max(sparse[kk][lefti], sparse[kk][righti - (1 << kk) + 1]);
                mg = imax;
                // jL
                {
                    int pj = Ps[jL];
                    int qj = Qs[jL];
                    int ljv = Ls[jL];
                    int rjv = Rs[jL];
                    int le = Math.min(pj - l, ljv);
                    int re = Math.min(r - qj, rjv);
                    mg = Math.max(mg, le + re);
                }
                // jR
                {
                    int pj = Ps[jR];
                    int qj = Qs[jR];
                    int ljv = Ls[jR];
                    int rjv = Rs[jR];
                    int le = Math.min(pj - l, ljv);
                    int re = Math.min(r - qj, rjv);
                    mg = Math.max(mg, le + re);
                }
            }
            answer.add(total_ones + mg);
        }
        return answer;
    }
}
# @lc code=end